File: mediator-pattern/chat-room/chat-room.js
// Mediator
var ChatRoom = /** @class */ (function () {
    function ChatRoom() {
    }
    ChatRoom.prototype.showMessage = function (user, msg) {
        console.log("[".concat(user, "] :  ").concat(msg));
    };
    return ChatRoom;
}());
// Colleague
var User = /** @class */ (function () {
    function User(name, chatroom) {
        this.name = name;
        this.chatroom = chatroom;
    }
    User.prototype.sendMessage = function (message) {
        this.chatroom.showMessage(this.name, message);
    };
    return User;
}());
// Usage
var chatroom = new ChatRoom();
var u1 = new User("Shreyash", chatroom);
var u2 = new User("Arpit", chatroom);
u1.sendMessage("Hola");
u2.sendMessage("Hi Shreyash, How are you ?");

File: mediator-pattern/chat-room/chat-room.rs
// Mediator trait
trait ChatRoomMediator {
    fn show_message(&self, user: &str, message: &str);
}

// Concrete Mediator
struct ChatRoom;

impl ChatRoomMediator for ChatRoom {
    fn show_message(&self, user: &str, message: &str) {
        println!("{}: {}", user, message);
    }
}

// Colleague
struct User<'a> {
    name: &'a str,
    chatroom: &'a dyn ChatRoomMediator,
}

impl<'a> User<'a> {
    fn new(name: &'a str, chatroom: &'a dyn ChatRoomMediator) -> Self {
        Self { name, chatroom }
    }

    fn send_message(&self, message: &str) {
        self.chatroom.show_message(self.name, message);
    }
}

fn main() {
    let chatroom1 = ChatRoom;

    let u1 = User::new("Shreyash", &chatroom1);
    let u2 = User::new("Gaurav", &chatroom1);

    u1.send_message("Hello Gaurav");
    u2.send_message("Hi Shreyash, How are you ?");
}
